### ISSUE TITLE
Error in google cloud run connecting with cloud SQL

### ISSUE THREAD

**Issue by iaarab** [prisma/prisma#1508]
The error message i get when running in docker with cloud run: 

`"\nInvalid `prisma.()` invocation in\n/usr/src/app/src/federations/b2bike/entities/service-pass/resolvers.js:6:76\n\n\n\nReason: Error occurred during query execution:\nConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timeout\n\nstack backtrace:\n   0: failure::backtrace::internal::InternalBacktrace::new\n   1: failure::backtrace::Backtrace::new\n   2: sql_query_connector::error::SqlError::into_connector_error\n   3: <std::future::GenFuture<T> as core::future::future::Future>::poll\n   4: <std::future::GenFuture<T> as core::future::future::Future>::poll\n   5: <std::future::GenFuture<T> as core::future::future::Future>::poll\n   6: <std::future::GenFuture<T> as core::future::future::Future>::poll\n   7: <std::future::GenFuture<T> as core::future::future::Future>::poll\n   8: std::future::poll_with_tls_context\n   9: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  10: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  11: std::panicking::try::do_call\n  12: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  13: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  14: <std::future::GenFuture<T> as core::future::future::Future>::poll\n  15: hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_catch\n  16: <hyper::server::conn::spawn_all::NewSvcTask<I,N,S,E,W> as core::future::future::Future>::poll\n  17: tokio::task::core::Core<T>::poll\n  18: std::panicking::try::do_call\n  19: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  20: tokio::task::harness::Harness<T,S>::poll\n  21: tokio::runtime::thread_pool::worker::GenerationGuard::run_task\n  22: tokio::runtime::thread_pool::worker::GenerationGuard::run\n  23: std::thread::local::LocalKey<T>::with\n  24: tokio::runtime::thread_pool::worker::Worker::run\n  25: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n  26: std::panicking::try::do_call\n  27: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  28: tokio::task::harness::Harness<T,S>::poll\n  29: tokio::runtime::blocking::pool::Inner::run\n  30: tokio::runtime::context::enter\n  31: std::sys_common::backtrace::__rust_begin_short_backtrace\n  32: std::panicking::try::do_call\n  33: __rust_maybe_catch_panic\n             at src/libpanic_unwind/lib.rs:78\n  34: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  35: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once\n             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942\n  36: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once\n             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942\n      std::sys_common::thread::start_thread\n             at src/libstd/sys_common/thread.rs:13\n      std::sys::unix::thread::Thread::new::thread_start\n             at src/libstd/sys/unix/thread.rs:79\n  37: start_thread\n  38: __clone\n) })\n"`

**Comment by pantharshit00** [comment]
@iaarab The error suggests that Prisma client was unable to reach your database server and the connection timed out. So, can you please make sure your database is up and running and also it is accessible in the cloud run environment?

**Comment by iaarab** [comment]
Hi Harshit. It is working locally so the problem is not in cloud SQL wich is available in the cloud run instance. I also use the database in the same app with the mysql node lib without problems. 

My datasource db URL looks like this: "mysql://user:password@public_ip:3306/database_name"

**Comment by dkozma** [comment]
A Cloud Run-compatible socketpath can be provided to Prisma, for instance:
```
postgres://USER:PASSWORD@localhost/DBNAME?host=/cloudsql/CLOUD_SQL_INSTANCE_CONNECTION_NAME
```

**Comment by iaarab** [comment]
@dkozma I tried with the following setup still getting the same error message.
```
datasource mysql {
  provider = "mysql"
  url = "mysql://USER:PASSWORD@PUBLIC_IP/DBNAME?host=/cloudsql/CLOUD_SQL_INSTANCE_CONNECTION_NAME
}
```

**Comment by dkozma** [comment]
Sorry for the delay, I need to verify it's working with `preview-021`. Also I haven't tested this with MySQL, so the connection string might be different. Will update once I've verified the latest release.

**Comment by pantharshit00** [comment]
Related: https://github.com/prisma/prisma-client-js/issues/437

I will wait on that before we can validate this.

**Comment by pimeys** [comment]
Hey,

Could you try the address in the format MySQL expects us to use? You can see a unit test where the format is visible [here](https://github.com/prisma/quaint/blob/master/src/connector/mysql.rs#L335). We try to follow MySQL and PostgreSQL preferences as they dictate in their manuals, and they differ here how they define the socket connections.

**Comment by pimeys** [comment]
And @nikolasburk this might need some refinement in our own documentation?

**Comment by dkozma** [comment]
Just wanted to verify that (PostgreSQL, at least) is working with `preview-021` using the following connection string:

```
postgres://USER:PASSWORD@localhost/DBNAME?host=/cloudsql/CLOUD_SQL_INSTANCE_CONNECTION_NAME
```

According to @pimeys ' post, the MySQL version should look something like:

```
mysql://USER:PASSWORD@localhost/DBNAME?socket=(/cloudsql/CLOUD_SQL_INSTANCE_CONNECTION_NAME)
```

**Comment by ludusrusso** [comment]
Did you solve this issue? I have the same issue connecting to cloud sql from cloud run. I've tried to connect directly with db public ip and it works correctly, but using `postgres://USER:PASSWORD@localhost/DBNAME?host=/cloudsql/CLOUD_SQL_INSTANCE_CONNECTION_NAME` it does not work!

**Comment by Sytten** [comment]
@ludusrusso Its working perfectly fine in our production environment. Check:
1. If the Cloud SQL Admin API is enabled (that one was the problem for me)
2. That you added the right instance in your Cloud Run configuration
3. Make sure your password doesn't contain weird characters (not sure what charset prisma accepts but its not the same as standard postgres, I had trouble with that too)

**Comment by dkozma** [comment]
Just to confirm it's been working OK for us as well (with the same connection string) in multiple Cloud Run environments (currently running `2.0.0-beta.5`).

**Comment by ludusrusso** [comment]
@Sytten thanks, I enabled the Cloud SQL Admin API and now it works perfectly!

**Comment by diegocouto** [comment]
Hi everyone,

Were any of you able to set up a socket connection when using MySQL as the provider? I've been trying to do so but I keep seeing the error:

```
"extensions": {
  "code": "INTERNAL_SERVER_ERROR",
  "exception": {
    "code": "P1001",
    "meta": {
      "database_host": "localhost",
      "database_port": 3306
    },
    "stacktrace": [
      "Error: ",
      "Invalid `prisma.user.findOne()` invocation in",
      "/usr/src/app/build/controllers/UserController.js:113:68",
      "",
      "",
      "  Can't reach database server at `localhost`:`3306`",
      "",
      "Please make sure your database server is running at `127.0.0.1`:`3306`.",
      "    at PrismaClientFetcher.request (/usr/src/app/node_modules/@prisma/client/runtime/index.js:1:148505)",
```

In order to limit the potential issues, I'm hardcoding the data source URL while testing this as such:

```
datasource db {
  provider = "mysql"
  url      = "mysql://user:psswd@localhost/db?host=/cloudsql/project:region:db"
}
```

To make sure the issue was not related to some missing permissions between Cloud Run and Cloud SQL, I've created a Knex client trying to connect to the same socket and it does connect successfully.

I'm not sure whether this could be an indicator or not but it's weird that the error message never mentions the socket, only the `IP:PORT`, making it look like it's always trying a TCP connection only.

Any thoughts? ☺️

**Comment by dkozma** [comment]
@diegocouto Have you tried the MySQL-specific connection string mentioned here (note `socket` and the parenthesis)? https://github.com/prisma/prisma/issues/1508#issuecomment-587728466

**Comment by pantharshit00** [comment]
I think this can be closed now as we have proper support for unix domain sockets and many people are successfully able to run their projects on google cloud run.

Please open a new issue if you are still facing this.

I have also merged @dkozma PR which fixes the docs for mysql.

**Comment by Nijinsha** [comment]
Hey trying to connect to cloud SQL database from app engine standard env. Getting this error.
`PrismaClientInitializationError2 [PrismaClientInitializationError]: Can't reach database server at 'hiddenIP':'3306'`

```
app.yaml has the url
DATABASE_URL: mysql://user:password@localhost/dbname?host=/cloudsql/projecy-stage:us-central1:dbname

beta_settings:
  cloud_sql_instances:project-id:us-central1:dbname
```

**Comment by pantharshit00** [comment]
Please open a new issue.
